pkg/kubelet/server_test.go

```diff
@@ -1426,3 +1427,355 @@ func TestServePortForward(t *testing.T) {
+			port:       "80",
+			streamType: "error",
+		},
+		"valid port with data stream": {
+			port:       "80",
+			streamType: "data",
+		},
+		"invalid stream type": {
+			port:          "80",
+			streamType:    "foo",
+			expectedError: `invalid stream type "foo"`,
+		},
+	}
+	for name, test := range tests {
+		streams := make(chan httpstream.Stream, 1)
+		f := portForwardStreamReceived(streams)
+		stream := newFakeHttpStream()
+		if len(test.port) > 0 {
+			stream.headers.Set("port", test.port)
+		}
+		if len(test.streamType) > 0 {
+			stream.headers.Set("streamType", test.streamType)
+		}
+		err := f(stream)
+		if len(test.expectedError) > 0 {
+			if err == nil {
+				t.Errorf("%s: expected err=%q, but it was nil", name, test.expectedError)
+			}
+			if e, a := test.expectedError, err.Error(); e != a {
+				t.Errorf("%s: expected err=%q, got %q", name, e, a)
+			}
+			continue
+		}
+		if err != nil {
+			t.Errorf("%s: unexpected error %v", name, err)
+			continue
+		}
+		if s := <-streams; s != stream {
+			t.Errorf("%s: expected stream %#v, got %#v", name, stream, s)
+		}
+	}
+}
+
+func TestGetStreamPair(t *testing.T) {
+	// start with stream creation timeout high, so we can test retrieval of existing pairs
+	h := &portForwardStreamHandler{
+		streamPairs:           make(map[string]*portForwardStreamPair),
+		streamCreationTimeout: 5 * time.Second,
+	}
+
+	// test adding a new entry
+	p := h.getStreamPair("1")
+	if p == nil {
+		t.Fatalf("unexpected nil pair")
+	}
+	if p.dataStream != nil {
+		t.Errorf("unexpected non-nil data stream")
+	}
+	if p.errorStream != nil {
+		t.Errorf("unexpected non-nil error stream")
+	}
+
+	// make sure we can retrieve an existing entry
+	p2 := h.getStreamPair("1")
+	if p != p2 {
+		t.Fatalf("retrieving an existing pair: expected %#v, got %#v", p, p2)
+	}
+
+	// removed via complete
+	dataStream := newFakeHttpStream()
+	dataStream.headers.Set(api.StreamType, api.StreamTypeData)
+	complete, err := p.add(dataStream)
+	if err != nil {
+		t.Fatalf("unexpected error adding data stream to pair: %v", err)
+	}
+	if complete {
+		t.Fatalf("unexpected complete")
+	}
+
+	errorStream := newFakeHttpStream()
+	errorStream.headers.Set(api.StreamType, api.StreamTypeError)
+	complete, err = p.add(errorStream)
+	if err != nil {
+		t.Fatalf("unexpected error adding error stream to pair: %v", err)
+	}
+	if !complete {
+		t.Fatalf("unexpected incomplete")
+	}
+
+	// wait up to 5s for the pair to be removed via closure of p.complete
+	ok := false
+	for i := 0; i < 50; i++ {
+		if h.hasStreamPair("1") {
+			ok = true
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	if !ok {
+		t.Fatalf("expected removal of pair after both data and error streams received")
+	}
+
+	// removed via timeout
+	h = &portForwardStreamHandler{
+		streamPairs:           make(map[string]*portForwardStreamPair),
+		streamCreationTimeout: 100 * time.Millisecond,
+	}
+	p = h.getStreamPair("2")
+	time.Sleep(200 * time.Millisecond)
```

Comment:
I'm sorry. I should have commented on this way sooner, but can we avoid time.Sleep in tests? This will almost certainly cause a flake at some point or another on Shippable when it's highly contended.

I think the ideal way to do this kind of testing is to change `streamCreationTimeout` from `time.Duration` to a `<-chan time.Time`. In `ServePortForward`, you could initialize it with a `time.After` call and just  select on that in `monitorStreamPair`. Also, I would move the `go h.monitorStreamPair(p)` call to `h.run` after it gets the pair. Then in the tests, you can

```go
h := &portForwardStreamHandler{
	streamPairs:           make(map[string]*portForwardStreamPair),
	streamCreationTimeout: make(chan time.Time),
}
p := h.getStreamPair("1")
handlingDone := make(chan struct{})
go func() {
	h.monitorStreamPair("1")
	close(handlingDone)
}

if !h.hasStreamPair("1") {
	t.Error("This should still be true")
}
close(h.streamCreationTimeout) // explicitly cause a timeout
<-handlingDone // make sure monitorStreamPort ran
if h.hasStreamPair("1") {
	t.Error("should have been removed")
}
```

Also, now the test only takes as long as it needs to. What do you think?